### DISCUSSION TITLE
Thoughts about Genre filter

### DISCUSSION THREAD

**Discussion by ghost** [SoongNoonien/mpdevil#109] [Ideas]
Firstly just wanted to say how much I'm enjoying V2 of Plattenalbum, the UX (cover art auto resize etc) is very slick and a pleasure to use plus the fuzzy search works great.

I've always had mixed views on Genre primarily because they are subjective to the point of being almost useless to my mind especially so when multiple Genre are applied to one album, however, I and possibly others have gone to great lengths to make our collections Genre meaningful... albeit if only to us.

V1 (mpdevil) had an option to display genre which worked but I always felt it spoiled the clean UI a bit when enabled and I've been pondering if there was a less intrusive option for those that might appreciate a Genre filter. I'm no developer and hence have no idea how hard / impractical my suggestion may be, but I wondered if it were possible to extend the current Artist search (thanks for implementing this btw) such that by making a search prefixed with `<Genre>` i.e. `<Blues>` for example would return a list of artists within that Genre ? Or maybe a keyboard shortcut would be simpler ?

Users with some developer skills may have some better ideas here ?

EDIT
Forgot code tags

**Comment by SoongNoonien** [comment]
> I've always had mixed views on Genre primarily because they are subjective to the point of being almost useless to my mind especially so when multiple Genre are applied to one album, however, I and possibly others have gone to great lengths to make our collections Genre meaningful... albeit if only to us.

Been there as well...

> V1 (mpdevil) had an option to display genre which worked but I always felt it spoiled the clean UI a bit when enabled and I've been pondering if there was a less intrusive option for those that might appreciate a Genre filter. I'm no developer and hence have no idea how hard / impractical my suggestion may be, but I wondered if it were possible to extend the current Artist search (thanks for implementing this btw) such that by making a search prefixed with <Genre> i.e. <Blues> for example would return a list of artists within that Genre ? Or maybe a keyboard shortcut would be simpler ?

The more important problem here is again that mpd doesn’t allow for searching anything other than songs. In other words it is not possible right now to ask mpd "Hey, give me a list of artists that have songs of a genre containing `searchstring` as a substring but don't worry about case or diacritics". See for reference https://github.com/MusicPlayerDaemon/MPD/issues/494. I'd say at least case insensitivity is a must have for a search feature. For example searching for "rock" one would expect results for "Rock" as well as for "Bluesrock". Right now only "Bluesrock" would be found by mpd. Of course we could work around that by fetching a list of genres and the respective artists within that genre from mpd and search them ourself. But I'm not a fan of this solution as this can potentially be a very long list that needs to be transmitted in its entirety over the network protocol just to be searched and thrown away for the most part.

**Reply by ghost** [reply]
I have no desire to see Plattenalbum become bogged down with client side code. I suppose I was just hoping the mpd workaround you implemented for the Artist list search could be extended to include Genre but I can see how this might make the list too long and result in unwanted lag. I also understand the risks to Plattenalbum of allowing the original goals to be compromised by feature creep, however, in the absence of the required work being done in mpd and the changes made in GTK4 we have ended up with a regression in features over V1.

The original proposal to add the required features in mpd dates back to 2019 and appears stalled I'm not hopeful of this changing anytime soon. I wish I had the skill set to undertake that work as I would happily commit the time to it.

**Reply by SoongNoonien** [reply]
> I suppose I was just hoping the mpd workaround you implemented for the Artist list search could be extended to include Genre but I can see how this might make the list too long and result in unwanted lag.

The nice thing about the workaround for the artist search is that the list of artists needs to be fetch form mpd anyway to be able to populate the player.

> I also understand the risks to Plattenalbum of allowing the original goals to be compromised by feature creep, however, in the absence of the required work being done in mpd and the changes made in GTK4 we have ended up with a regression in features over V1.

I get your point. The biggest problem with GTK4 regarding the search feature was the lack of an interactive search locally in the artist list. But I think the new artist search is better that the old one.

> The original proposal to add the required features in mpd dates back to 2019 and appears stalled I'm not hopeful of this changing anytime soon.

I afraid this might be true.

> I wish I had the skill set to undertake that work as I would happily commit the time to it.

Sadly, as pointed out in the bug report it can't be easily implemented.

**Reply by ghost** [reply]
> I think the new artist search is better that the old one.

Agreed it really does work very well!

I guess returning a list of genre wouldn't be any more load than the artists list and could be pulled up with a keyboard shortcut to replace the artist list ? The issue then is the number of albums which would be returned when one is selected. The maximum number of albums I have under a single artist is 67 whereas the maximum number of albums under a single genre is 5,482..... Ouch!

I will take a look at the mpd code but I suspect it's not a good project for a noob to cut his teeth on :-)
